Fix check mode

vcrhonek opened the pull request:
Fix the role if executed in check mode without having been executed in normal mode before.
https://github.com/linux-system-roles/timesync/issues/51

mlichvar commented:
The changes look good to me.

If it's not too much work, I think the switch to the package_facts would be better in a separate commit. Also, please consider using simpler commit subjects using only one sentence and moving the rest to the commit message, so the one-line git log looks nice.

The failure on RHEL8 looks like a selinux issue. I'll file a bug if it's something new.

pcahyna commented:
@mlichvar I believe it is https://bugzilla.redhat.com/show_bug.cgi?id=1668667 which should be fixed in the later snapshots (the rhel-x images).

mlichvar commented:
It seems there is another selinux issues that breaks the L2 transport and is not caught by our tests.
https://bugzilla.redhat.com/show_bug.cgi?id=1759214

I'll submit a PR to improve the test.

vcrhonek commented:
> The changes look good to me.

Thanks for review.

> If it's not too much work, I think the switch to the package_facts would be better in a separate
> commit.

Fixed. That particular commit also incorporates PR#49, because it touches that part of code anyway.

> Also, please consider using simpler commit subjects using only one sentence and moving
> the rest to the commit message, so the one-line git log looks nice.

Fixed.

pcahyna commented on tasks/main.yml:
```diff
@@ -54,6 +54,11 @@
+- name: Gather package facts
+  package_facts:
+    manager: auto
+  register: _timesync_package_facts
```
You don't need to register the output from a fact module.

vcrhonek replied:
I did that intentionally because otherwise the facts could change, right? We discussed it elsewhere in the past. But I see that this might be a different situation and register is not necessary.

vcrhonek replied:
Fixed.

pcahyna commented on tasks/main.yml:
```diff
@@ -68,24 +73,20 @@
-  check_mode: false
-  when: timesync_mode != 2 and timesync_ntp_provider == 'chrony'
+  set_fact:
+    timesync_chrony_version: "{{ _timesync_package_facts.ansible_facts.packages['chrony'][0].version|int }}"
```
just use ansible_facts.packages....
https://docs.ansible.com/ansible/latest/reference_appendices/common_return_values.html#ansible-facts

vcrhonek replied:
Fixed.

pcahyna commented on tasks/main.yml:
```diff
@@ -186,53 +198,67 @@
   service:
     name: ntpdate
     state: stopped
     enabled: false
```
It might just work in check mode, and even if it does not, ignore_errors below should make the "when" unneeded.

vcrhonek replied:
Correct, fixed.

pcahyna commented on tasks/main.yml:
```diff
@@ -68,24 +73,20 @@
-  changed_when: false
-  check_mode: false
-  when: timesync_mode != 2 and timesync_ntp_provider == 'chrony'
+  set_fact:
```
Please do not set a fact, use vars/main.yml and lazy evaulation, or set it on a block similarly to a (let) construct in LISP (which would have the advantage of making the variable really local to the block).

vcrhonek replied:
Fixed.

mlichvar commented:
The changes look good to me, but I'm not familiar with the lazy evaluation feature of ansible.  @pcahyna, can you please review the latest changes?

Please squash commits that fix previous changes if possible. (Don't squash all commits.)
- "Remove unnecessary 'when' conditions if task ignores failed commands" looks like a fix for "Fix role in check mode"
- "Don't register package facts" looks like a fix for "Use package_facts to determine chrony/ntp version".

vcrhonek commented:
I just rebased the pull request atop of the current master, no other changes done.

pcahyna commented:
> The changes look good to me, but I'm not familiar with the lazy evaluation feature of ansible. 

@mlichvar , I now wrote a longer explanation of lazy evaluation, here: https://github.com/linux-system-roles/storage/pull/81#discussion_r406069895
